src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -287,29 +295,9 @@ private static string GetPreferredUsernameFromIdToken(bool isAdfsAuthority, IdTo
 
             tokenCacheItems = FilterByHomeAccountTenantOrAssertion(requestParams, tokenCacheItems);
             tokenCacheItems = FilterByTokenType(requestParams, tokenCacheItems);
-
-            // no match found after initial filtering
-            if (!tokenCacheItems.Any())
-            {
-                logger.Info("No matching entry found for user or assertion");
-                return null;
-            }
-
-            if (logger.IsLoggingEnabled(LogLevel.Info))
-            {
-                logger.Info("Matching entry count - " + tokenCacheItems.Count());
-            }
-
             tokenCacheItems = FilterByScopes(requestParams, tokenCacheItems);
             tokenCacheItems = await FilterByEnvironmentAsync(requestParams, tokenCacheItems).ConfigureAwait(false);
 
-            // no match
-            if (!tokenCacheItems.Any())
```

Comment:
You've introduced a bug here. In most cases, there will be 0 tokens remaining after all these filters. But instead of stopping, you check GetSingleResult, which will fail with an exception stating that there are 1+ tokens and MSAL cannot decide.

Reply:
Good catch. This is why unit tests failed, but the error message was not accurate. 

If I add this check, could it happen that when line 307 executes, there are tokens in the cache, so we continue; then when we enter into `GetSingleResult`, another thread removes those tokens, and we end up with empty cache and the same error?

Similar point is that we call `.Any`, `.Count`, `.First` which apply all the predicates on the `IEnumerable` a few times in this search process. But other threads could update the cache in between these calls. Would it be better to have `IList<MsalAccessTokenCacheItem> tokenCacheItems = GetAllAccessTokensWithNoLocks(true).ToList();` above, so that we query a stable list?

There would still be a general perf benefit, since `HasTokensNoLocks` would still get an `IEnumerable` from `GetAllAccessTokens`.

Reply:
Done.